datafusion/functions-aggregate/src/count.rs

```diff
@@ -752,10 +761,245 @@ impl Accumulator for DistinctCountAccumulator {
     }
 }
 
+/// GroupsAccumulator for COUNT DISTINCT operations
+#[derive(Debug)]
+pub struct DistinctCountGroupsAccumulator {
+    /// One HashSet per group to track distinct values
+    distinct_sets: Vec<HashSet<ScalarValue, RandomState>>,
```

Comment:
I wonder if a single `HashSet<(u64, ScalarValue), RandomState>>` (i.e. also index by group id). might be faster?